src/modules/job-manager/jobtap.c

```diff
@@ -2078,6 +2111,54 @@ int flux_jobtap_event_post_pack (flux_plugin_t *p,
     return rc;
 }
 
+/* RFC 21 jobspec-update event keys must start with "attributes."
+ * Reject update events with keys that violate the RFC.
+ */
+static bool validate_jobspec_updates (json_t *o)
+{
+    const char *key;
+    json_t *entry;
+    json_object_foreach (o, key, entry) {
+        if (!strstarts (key, "attributes."))
+            return false;
+    }
+    return true;
+}
+
+int flux_jobtap_jobspec_update_pack (flux_plugin_t *p, const char *fmt, ...)
+{
+    int rc = -1;
+    int saved_errno;
+    va_list ap;
+    struct jobtap *jobtap;
+    json_t *o = NULL;
+    json_error_t error;
+
+    if (!p || !(jobtap = flux_plugin_aux_get (p, "flux::jobtap"))) {
+        errno = EINVAL;
+        return -1;
+    }
+    va_start (ap, fmt);
+    o = json_vpack_ex (&error, 0, fmt, ap);
+    va_end (ap);
+    if (!validate_jobspec_updates (o)) {
+        errno = EINVAL;
+        goto out;
+    }
```

Comment:
Handle `json_vpack_ex()` failure?

Reply:
Oh, good catch. Forgot to add that back after a refactor 🤦‍♂️